File: js/node.js
function Node() {
}

function BinaryNode(left, op, right) {
  Node.call(this);
  this.left = left;
  this.op = op;
  this.right = right;
}

BinaryNode.prototype = Object.create(Node.prototype);

BinaryNode.prototype.accept = function(visitor) {
  return visitor.visitBinary(this);
}

function CallNode(fn, args) {
  Node.call(this);
  this.fn = fn;
  this.args = args;
}

CallNode.prototype = Object.create(Node.prototype);

CallNode.prototype.accept = function(visitor) {
  return visitor.visitCall(this);
}

function NumberNode(value) {
  Node.call(this);
  this.value = value;
}

NumberNode.prototype = Object.create(Node.prototype);

NumberNode.prototype.accept = function(visitor) {
  return visitor.visitNumber(this);
}

function StringNode(value) {
  Node.call(this);
  this.value = value;
}

StringNode.prototype = Object.create(Node.prototype);

StringNode.prototype.accept = function(visitor) {
  return visitor.visitString(this);
}

function VariableNode(name) {
  Node.call(this);
  this.name = name;
}

VariableNode.prototype = Object.create(Node.prototype);

VariableNode.prototype.accept = function(visitor) {
  return visitor.visitVariable(this);
}

File: python/node.py
class Node:

class BinaryNode(Node):
  __init__(self, left, op, right):
    self.left = left
    self.op = op
    self.right = right

  def accept(self, visitor):
    return visitor.visitBinary(this)


class CallNode(Node):
  __init__(self, fn, args):
    self.fn = fn
    self.args = args

  def accept(visitor):
    return visitor.visitCall(self)


class NumberNode(Node):
  __init__(value):
    self.value = value

  def accept(self, visitor):
    return visitor.visitNumber(self)


class StringNode(Node):
  __init__(value):
    self.value = value

  def accept(self, visitor):
    return visitor.visitString(self)


class StringNode(Node):
  __init__(name):
    self.name = name

  def accept(self, visitor):
    return visitor.visitVariable(self)
